### DISCUSSION TITLE
Trouble accessing includes from another project / building with includes in another project

### DISCUSSION THREAD

**Discussion by hexratcc** [premake/premake-core#1919] [Q&A] [answered]
Hey, I have a workspace containing two projects: Engine (ConsoleApp) and Renderer (SharedLib) (+ GLFW and Glad) and I'm having trouble including a library inside the Renderer project and including the Renderer itself in the engine (If I try to compile just the Renderer I get no errors, but when I include the Renderer inside Engine I get a "cannot open source file" error). 

Project structure: 
```
Workspace
└─ Engine 
└─ Renderer
   └─ GLFW
   └─ Glad 
   └─ Vulkan 
   └─ glm 
```

I can access and call stuff located inside the Renderer, my issues arise whenever I try to do so **and** add ie. a GLFW #include somewhere in the Renderer. 

I have taken the liberty of creating a [repository ](https://github.com/Goubermouche/Engine)that recreates my problem. Note that if you plan on installing and compiling it yourself you will have to clone the repository **and** its submodules (use --recursive). 

Thanks in advance 😄.

**Comment by nickclark2016** [comment, marked as answer]
Are you exposing GLFW in a header file in Renderer? If so, then any dependencies of Renderer need to have the GLFW include directories.

**Reply by hexratcc** [reply]
Not exactly sure what you mean - I include Renderer in the Engine premake file and add it to includedirs - is that not enough? I should also note that the issue I've specified occurs while including any of the dependencies (that being GLFW,  Glad, glm and VK).

**Reply by nickclark2016** [reply]
Since I can't see the `Renderer` project, I'm speaking in hypothetical here.  Lets say `Renderer` has a header file `Renderer.h` that includes `glfw3.h`.  If in engine you include `Renderer.h` in file `Engine.cpp`, `Engine.cpp` needs to know about both `Renderer.h` and `glfw3.h`, as includes are basically just glorified copy-pastes.

**Reply by hexratcc** [reply]
I see, what would be the best solution for this then? Just "copy" the includedirs from `Renderer`'s premake file to `Engine`' premake file and modify the paths so that they point to the library directories inside the `Renderer`? or is there perhaps a cleaner/more streamlined way of doing this in premake? For the sake of completeness I'm also sending a link to the [Renderer ](https://github.com/Goubermouche/Renderer)repo.

**Reply by nickclark2016** [reply]
There's no way of doing this solely in Premake.  I personally use a table at the top level premake to hold all the dependency info (include paths, lib directories, etc).  You could also look at using various design patterns such that you don't expose the external dependency headers to consumers.  Simple inheritance, pimpl, and facades would be common ways to do this.

**Reply by hexratcc** [reply]
I see, I'll check it out then. Thanks for you help!
